File: src/main/java/com/github/pauldambra/moduluschecker/chain/checks/ExceptionFiveModulusElevenCheck.java
package com.github.pauldambra.moduluschecker.chain.checks;

import com.github.pauldambra.moduluschecker.Account.BankAccount;
import com.github.pauldambra.moduluschecker.ModulusCheckParams;
import com.github.pauldambra.moduluschecker.SortCodeSubstitution;
import com.github.pauldambra.moduluschecker.valacdosFile.WeightRow;

import java.util.function.Function;

public final class ExceptionFiveModulusElevenCheck {

    private final SortCodeSubstitution sortCodeSubstitution;

    public ExceptionFiveModulusElevenCheck(SortCodeSubstitution sortCodeSubstitution) {
        this.sortCodeSubstitution = sortCodeSubstitution;
    }

    public Boolean check(ModulusCheckParams params, Function<ModulusCheckParams, WeightRow> rowSelector) {
        WeightRow selectedRow = rowSelector.apply(params);

        BankAccount bankAccount = sortCodeSubstitution.Apply(params.account);
        params = params.withAccount(bankAccount);

        int total = ModulusTotal.calculate(params.account, selectedRow.getWeights());
        int remainder = total % 11;

        int checkDigit = params.account.getNumberAt(BankAccount.G);

        if (remainder == 0 && checkDigit == 0) {
            return true;
        }

        if (remainder == 1) {
            return false;
        }

        return (11 - remainder) == checkDigit;
    }
}

File: src/main/java/com/github/pauldambra/moduluschecker/chain/checks/ExceptionFiveModulusElevenCheck.kt
package com.github.pauldambra.moduluschecker.chain.checks

import com.github.pauldambra.moduluschecker.ModulusCheckParams
import com.github.pauldambra.moduluschecker.SortCodeSubstitution
import com.github.pauldambra.moduluschecker.account.BankAccount
import com.github.pauldambra.moduluschecker.valacdosFile.WeightRow

class ExceptionFiveModulusElevenCheck(private val sortCodeSubstitution: SortCodeSubstitution) {

    fun check(params: ModulusCheckParams, weightRow: WeightRow): Boolean {

        val bankAccount = sortCodeSubstitution.apply(params.account)
        val updatedParams = params.copy(account = bankAccount)

        val total = ModulusTotal.calculate(updatedParams.account, weightRow.weights)
        val remainder = total % 11

        val checkDigit = updatedParams.account.getNumberAt(BankAccount.G)

        return if (remainder == 0 && checkDigit == 0) {
            true
        } else if (remainder == 1) {
            false
        } else {
            11 - remainder == checkDigit
        }

    }
}
